ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -544,4 +530,127 @@ private boolean containsName(String name) {
+          } catch (Exception e) {
+            String message = "Updating accounts failed because unexpected exception occurred when updating accountId="
+                + account.getId() + " accountName=" + account.getName();
+            // Do not depend on Helix to log, so log the error message here.
+            logger.error(message, e);
+            throw new IllegalStateException(message, e);
+          }
+        }
+        newRecord.setMapField(ACCOUNT_METADATA_MAP_KEY, accountMap);
+        return newRecord;
+      }
+    }
+
+    /**
+     * Save the state from zookeeper from before the update and after the update to disk.
+     * This will write the backup content string to the disk with the following file name format:
+     * {@code {yyyyMMdd}T{HHmmss}-{unique long}.bak}.
+     * If there are multiple files with the same timestamp, the unique long will prevent the file names from clashing.
+     * @param succeeded {@code true} iff the update succeeded.
+     */
+    void persistBackup(boolean succeeded) {
+      try {
+        String timestamp = LocalDateTime.now().format(TIMESTAMP_FORMATTER);
+        JSONObject backupContent = getBackupContent(succeeded);
+        try (BufferedWriter writer = getWriter(timestamp)) {
+          backupContent.write(writer);
+        }
+      } catch (JSONException | IOException e) {
+        logger.error("Could not write backup file", e);
+      }
+    }
+
+    /**
+     * @param timestamp the timestamp for the backup file
+     * @return a {@link BufferedWriter} for the backup file.
+     * @throws IOException if the file could not be created.
+     */
+    private BufferedWriter getWriter(String timestamp) throws IOException {
+      for (long n = 0; n < Long.MAX_VALUE; n++) {
+        Path filepath = backupDirPath.resolve(timestamp + SEP + n + SEP + BACKUP_EXT);
+        try {
+          return Files.newBufferedWriter(filepath, StandardOpenOption.CREATE_NEW);
+        } catch (FileAlreadyExistsException e) {
+          // retry with a new suffix.
+        }
+      }
+      throw new IOException("Could not create a unique file with timestamp " + timestamp);
+    }
+
+    /**
+     * @param succeeded {@code true} iff the update succeeded.
+     * @return a {@link JSONObject} containing the content for the local backup.
+     * @throws JSONException
+     */
+    private JSONObject getBackupContent(boolean succeeded) throws JSONException {
+      JSONObject backupContent = new JSONObject();
+      backupContent.put(SUCCEEDED_KEY, succeeded);
+      JSONArray previousStateArray = new JSONArray();
+      if (remoteAccountInfoMap != null) {
+        for (Account account : remoteAccountInfoMap.getAccounts()) {
+          // We want to include the snapshot version read from zookeeper, not snapshot version + 1
+          previousStateArray.put(account.toJson(false));
+        }
+      }
+      backupContent.put(PREVIOUS_STATE_KEY, previousStateArray);
```

Comment:
if we were to put this json into the account update tool, would this simply work? Or would we need to make some manual changes? Would prefer it to work without changes.

Reply:
If you cut and paste the "previousState" array from the backup to a new file, this will work. Would you prefer having the previousState array in its own file so that it can be used directly?

Reply:
I think this should be good enough - this also depends on how you choose to address [this](https://github.com/linkedin/ambry/pull/818#discussion_r155912457) comment.